### DISCUSSION TITLE
[rant] ESM and its end goals

### DISCUSSION THREAD

**Discussion by Maksims** [playcanvas/engine#6994] [General]
This is a **very opinionated rant** about ESM and its end actual goals and its effect on the rest of engine and developers.
Would love to ask the team: **what is the end goal of ESM**.

### IMHO:
1. Cutting off a couple of dozen kilobytes from a built engine (treeshaking), is neglectable when talking about 95%+ of web apps using way more for assets.
2. The burden of doing imports/exports, forces developers to always pay attention: "did I forget to import a thing?", during development need to remember to remove not used any more imports. Although not writing `pc.` everywhere - is nice (can overcome with 5 lines of code though).
3. Flexibility-wise, the new ESM system does not do anything extra in terms of scripting system vs the current (classic) system. It has no new features and actually has fewer features.
4. It pollutes API with "alternative-ways", which means github, forums, code examples and everywhere we will see a mixture of ways of achieving the same goal while using different ways. It is bad for the learning curve of an API, and it has long-term lasting effects.
5. A lot of opinionated and "emotional" phrasing is used throughout docs regarding ESM, which means it is somewhat emotionally driven. Personally, I believe devs should make their own emotional attachments and beliefs regarding how great or not a feature is. Like the docs tries to justify the whole ESM thing.
6. We do use engine v2+ outside of the Editor, with ESM, and we see no much difference with projects where we use v1. The only major difference is that we have to use more sophisticated bundlers now, instead of simple script concatenation + minification. And due to bundlers taking a long time, this cannot be used while developing, like we used to use concat before to improve local development loading times.
7. The move towards ESM also does already affect and touch other parts: how to add components to entities, and this is only the beginning. This leads to changes, that do not actually help developers make better stuff.

And most importantly, **it takes vast amounts of time** from the engine/editor team to implement it. While that time could have been spent on way more impactful features:
1. Improving Editor UX
2. Introducing new Engine features
3. Introducing new Editor tools
4. Polishing existing Editor tools
8. Expanding Assets workflows
9. Making more great demos and improving documentation

Basically, creating stuff, instead of changing existing well-working things without a clear "end goal" that would benefit everyone.

It took around a year, and it is not here still. And who knows how much more it will take. While that time is spent on it, other features are not developed for the developers.

### My wish is that the team would focus on features and improvements that actually *help developers create better stuff*!

**Comment by marklundin** [comment]
Converting this to a disussion

**Reply by Maksims** [reply]
People don't read Discussions on GitHub.

**Reply by LeXXik** [reply]
If the repository is using it, then they do, actually. Check Jolt repo, for example. Not every library has a forum to turn to and issues are not great for q&a and discussions.

**Comment by slimbuck** [comment]
Maybe start with refuting @willeastcott's original issue :)

https://github.com/playcanvas/engine/issues/4767

**Comment by LeXXik** [comment]
I don't usually join rant convos. Its not like the team will read it and say "Oh, right, lets scrap everything we did so far". Just wanted to note that ESM is kind of a new feature for both engine and Editor. Maybe it being useful/impactful is also subjective to the dev using it. I kind of understand its not everyone's cup of tea.

**Comment by marklundin** [comment]
I try not to get involved with these kind of posts, as they're pretty subjective. The [original proposal](https://github.com/playcanvas/engine/issues/4767) explains everything and it seems to be [a popular feature request](https://github.com/playcanvas/engine/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-heart-desc). You also made some [good points yourself](https://x.com/mrmaxm/status/1777271028365840650). 

Your other points are well noted though, and we'll continue to look at other improvements to the editor :D
